### ISSUE TITLE
Lens dependency suggestion

### ISSUE THREAD

**Issue by andrewthad** [Soostone/katip#7]
I'm not a user of this library, just someone who saw it on reddit and decided to take a look. So, this is really just a suggestion that you can feel free to disregard. I noticed that `katip` depends on `lens` and `lens-aeson`. From a cursory glance, I can see that `lens` is used for a traversal somewhere and for generating lenses with template haskell. I'm not sure where `lens-aeson` is being used. Anyway, if your  use of these two is light enough, you may want to consider avoiding their use in `katip` itself (and just writing the lenses you provide by hand instead of using TH).That would make it a lighter install for users who don't need `lens`, and it would require less upper bound bumping as `lens` gets updated.

**Comment by MichaelXavier** [comment]
We've done something similar with `uri-bytestring` where you can define your non-exported type alias for `Lens` so your haddocks still refer to `Lens` and not the really hhhard to follow functor alias that Lens actually is. I will probably give this a try and see how difficult it is because we wouldn't want the lens dependency to be a detractor from people adopting the library.

**Comment by MichaelXavier** [comment]
@andrewthad what say you of https://github.com/Soostone/katip/commit/307a2daab3cf8b8dc6062c799402a8e395b7efd3

A colleague told me about the microlens family of packages which seem purpose-built for this. They don't have any dependencies besides base and transformers and export lenses compatible with the lens package. Are there any objections this does not address? Both packages appear to be in stackage and have been for some time.

**Comment by mightybyte** [comment]
@MichaelXavier You might want to ask Edward Kmett (or maybe the #lens IRC channel) about the tradeoffs involved here.  I seem to remember asking him about this awhile back and he had some decent reasons for why you shouldn't use the smaller lens packages.

**Comment by andrewthad** [comment]
@MichaelXavier I've put of a branch that goes further and removes the microlens dependency as well: https://github.com/andrewthad/katip/commit/c30b2bcd4938688350344bb8cea27bba8eb67983

I don't know anything about the issue that @mightybyte raises. My thoughts on microlens for this use-case are that `katip` doesn't internally make enough use of lenses and traversals to justify the dependency. They only really start be become more concise when you've got structures that are somewhat deeply nested.

Also, one practical advantage of hand-writing the lenses instead of using `mkLenses` is that you get nicer docs. You can add haddocks to them, and you don't get the `forall a.`. I've added example haddocks to the ones I wrote, although these obviously would need to be changed because I don't actually know how to describe what they do.

**Comment by andrewthad** [comment]
Although, just to be clear, at some point, I definitely think that just sucking it up and taking the dependency hit is worth it. It if were 200 lenses I was looking at writing, I'd take the `microlens-th` dependency any day.

**Comment by MichaelXavier** [comment]
@mightybyte was kind enough to send along a transcript of a discussion he had with Ed Kmett about lenses. It sounds like the main points where smaller lens libraries fall apart *in favor of lens itself* are:

1. If you need prisms and isos.
2. If you need some of the typeclass based things like ALens.

These don't apply to us. I can see the utility you pointed out in forgoing TH to derive lenses to hand-document lenses, but we already document record fields as needed and I strain to think of a case where the lensed version of a field needed documentation beyond what the field had. Some libraries actually don't export the record field and only export a default constructor and lenses (such as amazonka) in which case individual documentation of lenses seems like a good idea.

With these points aside, we come to whether we make enough use of lenses and traversals to justify the dependency. This is where things get tricky because I think it sounds like a matter of taste. I don't think anyone gains any insight by seeing 10+ hand-written lens definitions in a row vs TH. Also, if the `microlens` library is guaranteed to overlap on dependencies with the end user (i.e. base and transformers), and the compile time is negligible, I don't understand the resource we are conserving here.  I want to be very clear I am not trying to dismiss your approach out of hand, but if there are any points I haven't considered please let me know. From my perspective, `microlens{-th}` removes noise from the codebase and makes a few of the internal operations more concise at virtually no cost to the user.

**Comment by andrewthad** [comment]
@MichaelXavier I think our personal preferences diverge a little bit here. I generally try to avoid TH in libraries. I do agree that `microlens` (more particularly `microlens-th`) does make some things a little more clear, so I totally understand your perspective, and I think it's reasonable to keep the `microlens` dependency for the reasons you mention.

As a side note, one thing about `katip` that makes my usual "no TH in libraries" preference less applicable is that it's a logging library. So, you basically have to have stuff like `liftLoc` which gives you a `template-haskell` dependency anyways.

**Comment by MichaelXavier** [comment]
@andrewthad sounds like we're in agreement on the major points. I think we come out ahead with this change, where non lens-using users should benefit. I don't see any outstanding issues on katip that need to make next release, so I'll probably be cutting a release on this pretty soon. Thank you for the suggestions!
